processor/resourcedetectionprocessor/internal/aws/ecs/metadata.go

```diff
@@ -14,13 +14,94 @@
 
 package ecs
 
-import "net/http"
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+)
 
-type HTTPClient interface {
-	Do(req *http.Request) (*http.Response, error)
+type metadataProvider interface {
+	fetchTask(tmde string) (*TaskMetaData, error)
+	fetchContainer(tmde string) (*Container, error)
 }
 
-type ecsMetadataProvider interface {
-	fetchTaskMetaData(tmde string) (*TaskMetaData, error)
-	fetchContainerMetaData(tmde string) (*Container, error)
+type TaskMetaData struct {
+	Cluster          string
+	LaunchType       string // TODO: Change to enum when defined in otel collector convent
+	TaskARN          string
+	Family           string
+	AvailabilityZone string
+	Containers       []Container
+}
+
+type Container struct {
+	DockerID     string `json:"DockerId"`
+	ContainerARN string
+	Type         string
+	KnownStatus  string
+	LogDriver    string
+	LogOptions   LogData
+}
+
+type LogData struct {
+	LogGroup string `json:"awslogs-group"`
+	Region   string `json:"awslogs-region"`
+	Stream   string `json:"awslogs-stream"`
+}
+
+type metadataClient struct {
+	client *http.Client
+}
+
+var _ metadataProvider = &metadataClient{}
+
+// Retrieves the metadata for a task running on Amazon ECS
+func (md *metadataClient) fetchTask(tmde string) (*TaskMetaData, error) {
+	ret, err := fetch(tmde+"/task", md, true)
+	if ret == nil {
+		return nil, err
+	}
+
+	return ret.(*TaskMetaData), err
+}
+
+// Retrieves the metadata for the Amazon ECS Container the collector is running on
+func (md *metadataClient) fetchContainer(tmde string) (*Container, error) {
+	ret, err := fetch(tmde, md, false)
+	if ret == nil {
+		return nil, err
+	}
+
+	return ret.(*Container), err
+}
+
+func fetch(tmde string, md *metadataClient, task bool) (tmdeResp interface{}, err error) {
+	req, err := http.NewRequest(http.MethodGet, tmde, nil)
+
+	if err != nil {
+		log.Printf("Received error constructing request to ECS Task Metadata Endpoint: %v", err)
+		return nil, err
+	}
+
+	resp, err := md.client.Do(req)
+	if err != nil {
+		log.Printf("Received error from ECS Task Metadata Endpoint: %v", err)
+		return nil, err
+	}
+
+	if task {
+		tmdeResp = &TaskMetaData{}
+	} else {
+		tmdeResp = &Container{}
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(tmdeResp)
+	defer resp.Body.Close()
+
+	if err != nil {
+		log.Printf("Encountered unexpected error reading response from ECS Task Metadata Endpoint: %v", err)
```

Comment:
You'll want to use zap log instance passed in now.

Reply:
Can send it in another PR. This PR is just improving some readability issues. Added a TODO.